### ISSUE TITLE
Setting the affinity of threads in threadpools etc

### ISSUE THREAD

**Issue by ArashPartow** [David-Haim/concurrencpp#112]
Is there a way to set the cpu affinity (aka core affinity) of threads for instances of thread pools?  
_(eg: pthread_setaffinity_np for posix environments)_

#### Common Use Cases

-  **Scenario 1:**  List of M CPUs for an N thread thread_pool, where M <= N.  Where threads are part of a given CPU set and can ping-pong only within the specified set of CPUs. 

-  **Scenario 2:**  List of N CPUs for an N thread thread_pool. This is where each thread is pinned to one core only and is not allowed to run on any other core.

**Note1:** The above denoted CPU lists could be passed to the thread_pool as a constructor parameter.
**Note2:** The terms CPU and core are synonymous.

**Comment by David-Haim** [comment]
Hi there!
Thank you for taking interest in concurrencpp.


It is not possible to set the cpu affinity for threads in the thread pool. 
The threadpool enqueuing algorithm tries to enqueue new tasks to idle threadpool threads. The library prefers that an idle thread starts running a new task as soon as possible, even if it means running on a "foreign" core. 
Setting the cpu affinity is good when there are exactly threads as the logical core count, and they never block, which is not always the case in concurrencpp. 

Also, I think that in today's concurrency world, setting the affinity of threads is frowned upon, and it's better to let the OS scheduler decide when and where to run different threads. 

Lastly, concurrencpp tries to hide away as much as possible low level concurrency concepts such as OS threads, while utilizing different ways and tricks to optimize the concurrency infrastructure so the end-developer won't have to. trust the library and its performance! :)

**Comment by ArashPartow** [comment]
@David-Haim - thanks for the response.

I would say there are many scenarios where one would benefit from setting thread affinities.

For example in **Scenario 1** from above, it would be very beneficial to define the set of cores that the threads in the thread pool should run on.

This would make sure that threads never run on cores which are designated for specific tasks/executions.  Another example is NUMA centric - threads executing work that will be based on memory from a specific NUMA will not execute on cores in another NUMA.

------------

wrt usage within concurrencpp, nothing should change from a current usage point of view. Rather perhaps an extra constructor for the threadpool type can be provided that accepts either a list of core ids (for Scenario 1) representing the CPU's and will set the thread affinities of each thread as it is being constructed to the given cpuset.

In short: current usage should work fine and will require no code change and those that one the threads to be set upon a given set of cores can use the new constructor.

**Comment by jeremyong** [comment]
I agree that on most general desktop and virtualized environments, usage of thread affinity is not a good idea - you open yourself up to priority inversion bugs, and performance isn't really a guaranteed outcome, given that you don't necessarily know what other applications are running concurrently on the system.

However, maybe for users such as the OP, it'd be useful to provide an example showing how this could be done in userspace. It seems relatively straightforward to create a couple custom executors, each creating threads masked for execution on a set of cores. Some use cases are systems such as game consoles or HPC environments, where the set of running applications is known deterministically, and the CPU topology is similarly well-understood and controlled.
